### PR TITLE
gadgets: Provide human readable version of error codes

### PR DESCRIPTION
**Pull request by burak-ok**
# gadgets: Provide human readable version of error codes

```
$ go run --exec "sudo -E" ./cmd/ig/... run t --verify-image=false --fields task,src,dst,retcode
INFO[0000] Experimental features enabled                
WARN[0000] Ignoring runtime "podman" with non-existent socketPath "/run/podman/podman.sock" 
WARN[0000] you set --verify-image=false, image will not be verified 
WARN[0000] you set --verify-image=false, image will not be verified 
TASK       SRC                      DST                   RETCODE            
wget       172.17.0.2:51382         1.1.1.1:80                               
wget       172.17.0.2:45834         1.1.1.1:443                              
wget       172.17.0.2:45846         1.1.1.1:443                              
wget       0.0.0.0:0                192.168.2.1:0         ECONNREFUSED       
wget       0.0.0.0:0                192.168.1.2:0         EHOSTUNREACH       
wget       172.17.0.2:37268         192.168.2.50:80       unknown error -512
```

1. the first 3 lines are from a succesfull connect.
2. forth line is where my router refused the connection
3. fifth is an address in a subnet which I don't have access to
4. sixth I canceled the wget command with `Ctrl+C`

Problems:
1. [Not problem of this PR] So for some of the error cases we seem to have lost the source ip address. I think it was because of https://github.com/inspektor-gadget/inspektor-gadget/pull/3069
2. So the `unknown error -512` is actually `ERESTARTSYS` https://elixir.bootlin.com/linux/latest/source/include/linux/errno.h#L14 . The comment mentions that it shouldn't be seen by user programs. Also the comment metions that it still can be seen by `ptrace`, which I tested - [yes it is seen and resolved](https://github.com/search?q=repo%3Astrace%2Fstrace%20ERESTARTSYS&type=code). Do we just accept the fact that we show this "to the user"?

### CODE REVIEW

File: pkg/operators/formatters/formatters.go [hunk 1]
```diff
@@ -230,6 +234,40 @@ var replacers = []replacer{
...
+				switch len(inBytes) {
+				default:
+					return nil
+				case 4:
+					errnoNumber, _ := in.Int32(data)
+					errnoName := unix.ErrnoName(syscall.Errno(-errnoNumber))
+					if errnoNumber != 0 && errnoName == "" {
+						errnoName = fmt.Sprintf("unknown error %d", errnoNumber)
+					}
+					errnoField.Set(data, []byte(errnoName))
```
Comment (mauriciovasquezbernal) [inline pkg/operators/formatters/formatters.go:264]: We now have `PutString()` to avoid the casting.

File: pkg/operators/formatters/formatters.go [hunk 2]
```diff
@@ -230,6 +234,40 @@ var replacers = []replacer{
...
+			errnoField, err := ds.AddField(outName, api.Kind_String)
+			if err != nil {
+				return nil, err
+			}
+
+			in.SetHidden(true, false)
+
+			return func(data datasource.Data) error {
+				inBytes := in.Get(data)
+				switch len(inBytes) {
```
Comment (mauriciovasquezbernal) [inline pkg/operators/formatters/formatters.go:255]: Perhaps better to use a switch on `in.Type()` and call the right getter according to it? (I suppose we only need to take care of int32 for now).

File: gadgets/trace_mount/go/program.go [hunk 3]
```diff
@@ -99,9 +99,9 @@ func gadgetInit() int {
 		dest, _ := destField.String(data)
 		fs, _ := fsField.String(data)
 		dataStr, _ := dataField.String(data)
-		ret, _ := retField.Int32(data)
+		retRaw, _ := retRawField.Int32(data)
 
-		callField.SetString(data, getCallStr(opRaw, src, dest, fs, flags, dataStr, ret))
+		callField.SetString(data, getCallStr(opRaw, src, dest, fs, flags, dataStr, retRaw))
```
Comment (mauriciovasquezbernal) [inline gadgets/trace_mount/go/program.go:104]: Why not to provide the human readable version here too?

**Review by mauriciovasquezbernal — COMMENTED** [review]
> So the unknown error -512 is actually ERESTARTSYS https://elixir.bootlin.com/linux/latest/source/include/linux/errno.h#L14 . The comment mentions that it shouldn't be seen by user programs. Also the comment metions that it still can be seen by ptrace, which I tested - [yes it is seen and resolved](https://github.com/search?q=repo%3Astrace%2Fstrace%20ERESTARTSYS&type=code). Do we just accept the fact that we show this "to the user"?

As gadgets can potentially hook into internal functions of the kernel (not only syscalls), I'd say it's fine to show those errors to. Is it too complicated to also translate them?

I did a quick pass and it looks very good, thanks for working on this.

Any reason not to use it on trace open as well? 

https://github.com/inspektor-gadget/inspektor-gadget/blob/e5703c3ba52852f989274d381db27c1ecef42e81/gadgets/trace_open/program.bpf.c#L31

**Comment by burak-ok** [comment]
> Any reason not to use it on trace open as well?

Reason was that it didn't come up when I searched for `ret` or `retcode` :D

File: gadgets/trace_mount/go/program.go (hunk 3 above)
Comment (burak-ok) [reply gadgets/trace_mount/go/program.go:104]: I think the priority of the operators is getting in the way. I tried to get the `ret` field, but it is apparently not known

**Review by mauriciovasquezbernal — COMMENTED** [review]
It looks very good. 

Just a blocker on the trace open gadget as it doesn't work fine. 

I also see the metadata file hasn't been updated in all gadgets, are you planning to do it? Ideally the formatter should be able to add these annotations to the new field. I tried to implement it on #3149, but I stopped it, perhaps it's time to explore it again.

File: gadgets/trace_open/program.bpf.c [hunk 4]
```diff
@@ -158,7 +158,7 @@ static __always_inline int trace_exit(struct syscall_trace_exit *ctx)
 	bpf_probe_read_user_str(&event->fname, sizeof(event->fname), ap->fname);
 	event->flags_raw = ap->flags;
 	event->mode_raw = ap->mode;
-	event->err = errval;
+	event->err_raw = errval;
```
[the remark this answers is not in the document]
Comment (burak-ok) [reply gadgets/trace_open/program.bpf.c:161]: I guess we could make the contract that it should be positive, which could be easily enforcible with having the type an unsigned int.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply gadgets/trace_open/program.bpf.c:161]: Having it positive makes sense to me.

**Comment by burak-ok** [comment]
> > > Any reason not to use it on trace open as well?
> > 
> > 
> > Reason was that it didn't come up when I searched for `ret` or `retcode` :D
> 
> Maybe this is the opportunity to uniform them all? Probably in another PR, if you prefer, but you already did the work of identifying them all and their differences (half of the work 😝)

You are right, I guess `errno` or `error` would be the best name. The rest of the fields is done in https://github.com/inspektor-gadget/inspektor-gadget/pull/3169 :D

**Review by mauriciovasquezbernal — APPROVED** [review]
It already LGTM. Just one comment about the annotations / template, up to you if you want to handle it here or not.

File: gadgets/trace_signal/gadget.yaml [hunk 5]
```diff
@@ -27,10 +27,9 @@ datasources:
       gid:
         annotations:
           template: uid
-      ret:
+      error_raw:
         annotations:
-          columns.width: 5
```
[the remark this answers is not in the document]
Comment (burak-ok) [reply gadgets/trace_signal/gadget.yaml:32]: I changed it to the `_raw` field -> This field is hidden anyway.

File: gadgets/trace_open/gadget.yaml [hunk 6]
```diff
@@ -37,11 +37,9 @@ datasources:
       mode:
         annotations:
           description: File access mode
-      err:
+      error_raw:
         annotations:
-          description: Error code
-          columns.width: 5
-          columns.alignment: right
+          columns.hidden: true
```
[the remark this answers is not in the document]
Comment (burak-ok) [reply gadgets/trace_open/gadget.yaml:42]: You mean for the `error_raw` variant and for `error` itself? What would be the default width for `error` as a string.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply gadgets/trace_open/gadget.yaml:42]: We don't care that much about the error field (as it's hidden by default). `error` will use the [global default which is 16.](https://github.com/inspektor-gadget/inspektor-gadget/blob/858faff47af7f0cde353f90a622c91e9da01ba15/pkg/datasource/columns.go#L210)

**Pull request merged**